src/v/cluster/data_migration_backend.cc

```diff
@@ -40,27 +54,753 @@ backend::backend(
+            auto new_shard = _shard_table.shard_for(ntp_req.ntp);
+            update_partition_shard(ntp_req.ntp, rwstate, new_shard);
+        } break;
+        case migrated_replica_status::done:
+            reply.actual_states.push_back(
+              {.ntp = ntp_req.ntp,
+               .migration = metadata.id,
+               .state = ntp_req.state});
+        }
+    }
 
     vlog(dm_log.debug, "node request reply: {}", reply);
     return ssx::now(std::move(reply));
 }
 
+void backend::update_partition_shard(
+  const model::ntp& ntp,
+  replica_work_state& rwstate,
+  std::optional<ss::shard_id> new_shard) //
+{
+    vlog(
+      dm_log.trace,
+      "for ntp {} for migration {} seeking state {} updating shard: {} => "
+      "{}",
+      ntp,
+      rwstate.migration_id,
+      rwstate.sought_state,
+      rwstate.shard,
+      new_shard);
+    if (new_shard != rwstate.shard) {
+        if (rwstate.shard) {
+            stop_partition_work(ntp, rwstate);
+        }
+        rwstate.shard = new_shard;
+        if (new_shard) {
+            start_partition_work(ntp, rwstate);
+        }
+    }
+}
+
+void backend::clear_tstate_belongings(
+  const model::topic_namespace& nt,
+  const topic_reconciliation_state& tstate) //
+{
+    const auto& partitions = tstate.outstanding_partitions;
+    for (const auto& [partition, nodes] : partitions) {
+        for (const model::node_id& node : nodes) {
+            auto ns_it = _nstates.find(node);
+            ns_it->second.erase({nt.ns, nt.tp, partition});
+            if (ns_it->second.empty()) {
+                _nodes_to_retry.erase(node);
+                _nstates.erase(ns_it);
+            }
+        }
+    }
+}
+
+void backend::drop_mrstate(mrstates_t::const_iterator rs_it) {
+    const auto& topics = rs_it->second.outstanding_topics;
+    for (const auto& [nt, tstate] : topics) {
+        clear_tstate_belongings(nt, tstate);
+        _work_states.erase(nt);
+        _topic_migration_map.erase(nt);
+    }
+    _mrstates.erase(rs_it);
+}
+
+ss::future<> backend::reconcile_topic(
+  const model::topic_namespace& nt,
+  topic_reconciliation_state& tstate,
+  id migration,
+  state sought_state,
+  bool schedule_local_work) //
+{
+    if (!schedule_local_work && !_is_coordinator) {
+        vlog(
+          dm_log.debug,
+          "not tracking topic {} transition towards state {} as part of "
+          "migration {}",
+          nt,
+          sought_state,
+          migration);
+        co_return;
+    }
+    vlog(
+      dm_log.debug,
+      "tracking topic {} transition towards state {} as part of "
+      "migration {}, schedule_local_work={}, _is_coordinator={}",
+      nt,
+      sought_state,
+      migration,
+      schedule_local_work,
+      _is_coordinator);
+    auto assignments = _topic_table.get_topic_assignments(nt);
+    if (!assignments) {
+        // topic looks disabled, we'll get a notification when it is enabled
+        // and retry to coordinate
```

Comment:
this comment is not valid, if the topic is disabled the assignment will still be there, it can only be missing if the topic was removed